x-pack/platform/plugins/shared/observability_ai_assistant/server/service/distributed_lock_manager/index.ts

```diff
@@ -0,0 +1,329 @@
+      }
+    } catch (e) {
+      if (
+        e instanceof errors.ResponseError &&
+        e.body?.error?.type === 'version_conflict_engine_exception'
+      ) {
+        this.logger.debug(`Lock "${this.lockId}" already held (version conflict)`);
+        return false;
+      }
+
+      this.logger.error(`Failed to acquire lock "${this.lockId}": ${e.message}`);
+      this.logger.debug(e);
+      return false;
+    }
+  }
+
+  /**
+   * Releases the lock by deleting the document with the given lockId and token
+   */
+  public async release(): Promise<boolean> {
+    try {
+      const res = await pRetry(
+        () => {
+          return this.esClient.deleteByQuery({
+            index: LOCKS_CONCRETE_INDEX_NAME,
+            query: {
+              // bool: { filter: [{ term: { _id: this.lockId } }, { term: { token: this.token } }] },
+              bool: { filter: [{ term: { _id: this.lockId } }, { term: { token: this.token } }] },
+            },
+            refresh: true,
+          });
+        },
+        { retries: 2 }
+      );
+
+      if (res.deleted === 0) {
+        this.logger.debug(
+          `Lock "${this.lockId}" with token = ${this.token} could not be released. Not found or already released.`
+        );
+        return false;
+      }
+
+      this.logger.debug(`Lock "${this.lockId}" with token = ${this.token} was released.`);
+      return true;
+    } catch (error) {
+      this.logger.error(`Failed to release lock "${this.lockId}": ${error.message}`);
+      this.logger.debug(error);
+      return false;
+    }
+  }
+
+  /**
+   * Releases a lock if it is expired.
+   */
+  public async releaseIfExpired(): Promise<boolean> {
+    const res = await this.esClient.deleteByQuery({
+      index: LOCKS_CONCRETE_INDEX_NAME,
+      query: {
+        bool: {
+          filter: [{ term: { _id: this.lockId } }, { range: { expiresAt: { lte: 'now' } } }],
+        },
+      },
+      refresh: true,
+    });
+    if (res.deleted === 0) {
+      this.logger.debug(`Lock "${this.lockId}" not found`);
+      return false;
+    }
+
+    this.logger.debug(`Lock "${this.lockId}" expired and was released.`);
+    return true;
+  }
+
+  /**
+   * Retrieves the lock document for a given lockId.
+   * If the lock is expired, it will be released.
+   */
+  public async get(): Promise<LockDocument | undefined> {
+    const result = await this.esClient.search<LockDocument>({
+      index: LOCKS_CONCRETE_INDEX_NAME,
+      query: {
+        bool: {
+          filter: [{ term: { _id: this.lockId } }, { range: { expiresAt: { gt: 'now' } } }],
+        },
+      },
+    });
+
+    const hits = result.hits.hits;
+    if (hits.length === 0) {
+      await this.releaseIfExpired();
+      return undefined;
+    }
+
+    return hits[0]._source;
+  }
+
+  public async waitForLock(options: AcquireOptions & pRetry.Options = {}): Promise<boolean> {
+    const { metadata, ttlMs, ...pRetryOptions } = options;
+
+    return pRetry(async () => {
```

Comment:
I'm not sure about this retrying until it is acquired. I think quite often the lock being there just means that the operation shouldn't run again. is it something we can push to the consumer?

Reply:
This is a helper that consumers can use. We don't call it out of the box.

Reply:
I've now exposed only `withLock` in the server. All the others are not exposed. Let's focus on that